File: docs/js/comparator.js
import { Feedback, Match, PriceRelation } from "./models.js";

export const valuesMatch = (guess, answer) => [...guess].some((value) => answer.has(value));

export function compare(guess, answer) {
  let price = PriceRelation.EQUAL;
  if (guess.price !== null && answer.price !== null) {
    price = guess.price === answer.price ? PriceRelation.EQUAL : answer.price < guess.price ? PriceRelation.LOWER : PriceRelation.HIGHER;
  }
  return new Feedback({
    tradeable: guess.tradeable === answer.tradeable ? Match.MATCH : Match.NO_MATCH,
    price,
    equipment_slot: guess.equipment_slot !== null && answer.equipment_slot !== null && guess.equipment_slot === answer.equipment_slot ? Match.MATCH : Match.NO_MATCH,
    region: valuesMatch(guess.regions, answer.regions) ? Match.MATCH : Match.NO_MATCH,
    category: valuesMatch(guess.categories, answer.categories) ? Match.MATCH : Match.NO_MATCH,
    source: valuesMatch(guess.sources, answer.sources) ? Match.MATCH : Match.NO_MATCH,
    name: guess.name === answer.name ? Match.MATCH : Match.NO_MATCH,
  });
}

File: cloggle/comparator.py
from .models import Item, Feedback, Match, PriceRelation


def values_match(
    guess: set[str],
    answer: set[str],
) -> bool:
    return bool(guess & answer)


def compare(guess: Item, answer: Item) -> Feedback:
# Price relation: if either price is missing, treat as EQUAL (no information)
    if guess.price is None or answer.price is None:
        price = PriceRelation.EQUAL
    else:
        if guess.price == answer.price:
            price = PriceRelation.EQUAL
        elif answer.price < guess.price:
            price = PriceRelation.LOWER
        else:
            price = PriceRelation.HIGHER

    return Feedback(
    tradeable=(
        Match.MATCH
        if guess.tradeable == answer.tradeable
        else Match.NO_MATCH
    ),

    price=price,

    equipment_slot=(
        Match.MATCH
        if (
            guess.equipment_slot is not None
            and answer.equipment_slot is not None
            and guess.equipment_slot == answer.equipment_slot
        )
        else Match.NO_MATCH
    ),

    region=(
        Match.MATCH
        if values_match(guess.regions, answer.regions)
        else Match.NO_MATCH
    ),

    category=(
        Match.MATCH
        if values_match(guess.categories, answer.categories)
        else Match.NO_MATCH
    ),

    source=(
        Match.MATCH
        if values_match(guess.sources, answer.sources)
        else Match.NO_MATCH
    ),

    name=(
        Match.MATCH
        if guess.name == answer.name
        else Match.NO_MATCH
    ),
)